### ISSUE TITLE
[FIRE-4513] reporting DROP and REJECT rules as unused.

### ISSUE THREAD

**Issue by mdklapwijk** [CISOfy/lynis#856]
**Describe the bug**
Test FIRE-4513 reports DROP and REJECT rules as being unused, but shouldn't this be a good thing?

**Version**
 - Debian 10
 - Lynis version 2.7.5

**Expected behavior**
Only report on non-blocking rules being unused.

**Output**
```
root@prx-01:~# lynis show details FIRE-4513
2020-03-17 14:16:06 Performing test ID FIRE-4513 (Check iptables for unused rules)
2020-03-17 14:16:06 Result: Found one or more possible unused rules
2020-03-17 14:16:06 Description: Unused rules can be a sign that the firewall rules aren't optimized or up-to-date
2020-03-17 14:16:06 Note: Sometimes rules aren't triggered but still in use. Keep this in mind before cleaning up rules.
2020-03-17 14:16:06 Output: iptables rule numbers: 3 6
2020-03-17 14:16:06 Suggestion: Check iptables rules to see which rules are currently not used [test:FIRE-4513] [details:-] [solution:-]
2020-03-17 14:16:06 Tip: iptables --list --numeric --line-numbers --verbose
2020-03-17 14:16:06 ===---------------------------------------------------------------===
```

```
root@prx-01:~# iptables --list --numeric --line-numbers --verbose
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination
1        1    96 ACCEPT     icmp --  *      *       0.0.0.0/0            0.0.0.0/0            /* 000 accept all icmp requests */
2        4   116 ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0            /* 001 accept all to lo interface */
3        0     0 REJECT     all  --  !lo    *       0.0.0.0/0            127.0.0.0/8          /* 002 reject local traffic not on loopback interface */ reject-with icmp-port-unreachable
4     4127  629K ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED /* 003 accept related established rules */
5        0     0 ACCEPT     tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            multiport dports 22 /* 004 allow inbound ssh */
6        0     0 DROP       all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* 999 drop all other requests */

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination
# Warning: iptables-legacy tables present, use iptables-legacy to see them
```

**Comment by rm-td** [comment]
Hi, I dont think your "bug" above is correct.
Far as I understand the rule, is that is is looking for 0 packets/bytes processed by any particular line. This would be why the remediation suggests "iptables --list --numeric --verbose" and look for unused lines.

Your lines 5 and 6 will, in all likelihood NEVER be hit as line 4 blindly accepts all packets. In which case the packet count won't increment therefore the rule will always sit as "unused" which is what FIRE-4513 reports on.
I'd suggest your firewall rules are not totally correct.

Check out this example from one of my machines you will see that each rule, including DROP is populated as its possible for a packet to hit any one of the rulesets. Your example above is impossible.

```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
48356   27M ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
  991 46449 DROP       all  --  eth0   *       0.0.0.0/0            0.0.0.0/0            match-set geoblock src
  147  8746 DROP       all  --  eth0   *       0.0.0.0/0            0.0.0.0/0            match-set dynblock src
    5   200 DROP       all  --  eth0   *       0.0.0.0/0            0.0.0.0/0            match-set hostblock_ip src
  554 28276 DROP       tcp  --  eth0   *       0.0.0.0/0            0.0.0.0/0            tcp multiport dports 23,445
 2374  839K DROP       udp  --  eth0   *       0.0.0.0/0            0.0.0.0/0            udp multiport dports 68
 1906  135K ACCEPT     all  --  lo     *       0.0.0.0/0            0.0.0.0/0
    2  1280 ACCEPT     tcp  --  *      *       0.0.0.0/0            192.168.100.32       tcp match-set admin_port src multiport dports 22,6088 state NEW
   91  5320 ACCEPT     tcp  --  *      *       0.0.0.0/0            192.168.100.33       tcp multiport dports 80,443 state NEW
   85  3883 RULE_8     all  --  *      *       0.0.0.0/0            0.0.0.0/0

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
34324   61M ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state RELATED,ESTABLISHED
 1906  135K ACCEPT     all  --  *      lo      0.0.0.0/0            0.0.0.0/0
 2359  196K ACCEPT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            state NEW

Chain RULE_8 (1 references)
 pkts bytes target     prot opt in     out     source               destination
   85  3883 LOG        all  --  *      *       0.0.0.0/0            0.0.0.0/0            LOG flags 0 level 6 prefix "firewall -- DENY "

```

**Comment by mdklapwijk** [comment]
Line 4 is for ESTABLISHED, 5 and 6 do get triggered/used when I scan the ports, after which the FIRE-4513 only reports line 3 as being unused.

**Comment by rm-td** [comment]
"5 and 6 do get triggered when I scan the ports" - OK so FIRE-4513 is working to spec - on resetting your firewall the packet count will be zero until something triggers that line. Your reply is contrary to first report. Per above, I'd say its why the resolution is to run the iptables command and inspect the empty lines. It also clearly says "Note: Sometimes rules aren't triggered but still in use." Thats for a competent admin to determine.

"line 3 is unused" - again review what FIRE-4513 is telling you - "one or more possible unused rules"

Have a think about this - can you suggest in what real-world circumstances you can engineer traffic to loopback to be sucessfully transported over anything but a local interface? Think you'll find thats Why the rule is sitting at 0 packets logged because nothing will hit it, and in turn Why an unused rule is reported?

**Comment by mdklapwijk** [comment]
I raised this ticket, because the test is reporting REJECT and DROP rules as being unused, which in my opinion, should not be listed as a warning for unused rules. Because if everything is setup and behaving correctly, those rules should have a zero count.

Having your explicit drop/reject rules on a local host firewall have non-zero counts should be a reason for concern, as there would be 'something' on the host or in your network violating your policies.

Not debating the possibility/validity of rule 3, it is something that is being done quite often to prevent some forms of spoofing.

So I created this ticket to raise the following question: Should test FIRE-4513 be warning about firewall rules with a blocking state having a zero-count?
